lib/github/kv.rb

```diff
@@ -252,6 +253,50 @@ def setnx(key, value, expires: nil)
       }
     end
 
+    # increment :: String, Integer, expires: Time? -> Integer
+    #
+    # Increment the key's value by an amount.
+    #
+    # key     - The key to increment.
+    # amount  - The amount to increment the key's value by.
+    # expires - When the key should expire.
+    #
+    # Returns the key's value after incrementing.
+    def increment(key, amount: 1, expires: nil)
+      sql = GitHub::SQL.run(<<-SQL, key: key, amount: amount, now: now, expires: expires || GitHub::SQL::NULL, connection: connection)
+        INSERT INTO key_values (`key`, `value`, `created_at`, `updated_at`, `expires_at`)
+        VALUES(:key, :amount, :now, :now, :expires)
+        ON DUPLICATE KEY UPDATE
+          `value`=IF(
+            concat('',`value`*1) = `value`,
+            LAST_INSERT_ID(IF(
+              `expires_at` IS NULL OR `expires_at`>=NOW(),
+              `value`+:amount,
+              :amount
+            )),
+            `value`
+          ),
+          `updated_at`=IF(
+            concat('',`value`*1) = `value`,
+            :now,
+            `updated_at`
+          ),
+          `expires_at`=IF(
+            concat('',`value`*1) = `value`,
+            :expires,
+            `expires_at`
+          )
+      SQL
+
+      if sql.last_insert_id == 0
```

Comment:
My original plan was to use `sql.affected_rows` _which_ should be `0` when the update doesn't change any values. However I couldn't get that working...

> With ON DUPLICATE KEY UPDATE, the affected-rows value ..., and 0 if an existing row is set to its current values.

For some reason, even though I set them all to the current value it would still return 1 no matter what.

I did however notice that when an update occurs and no data is modified the `sql.last_insert_id` is simply set to 0. Through this I'm able to determine if the caller attempted to update a key who's value was non-numeric.

Reply:
Perhaps we pass the `CLIENT_FOUND_ROWS` flag to the C API? I'm checking with the DB folks to see what they might know.

> If you specify the CLIENT_FOUND_ROWS flag to the mysql_real_connect() C API function when connecting to mysqld, the affected-rows value is 1 (not 0) if an existing row is set to its current values.

Reply:
It looks like `CLIENT_FOUND_ROWS` [is set in keiko](https://github.slack.com/archives/C0FHZDN13/p1564783225007100). Looks like we'll have to stick with the `last_insert_id` workaround. 💥

Reply:
> It looks like CLIENT_FOUND_ROWS is set in keiko. Looks like we'll have to stick with the last_insert_id workaround. 💥

Is there any way for us to detect this from the library code? It would be nice if we could fall back to doing a second query if the keiko configuration ever changed.

Reply:
We're not relying on the keiko configuration. My workaround of using `last_insert_id` is not affected by `CLIENT_FOUND_ROWS`.

However if we wanted to detect if the `CLIENT_FOUND_ROWS` is set we could do something along the lines of:

```
flags = ActiveRecord::Base.connection.raw_connection.query_options[:flags]
check = Mysql2::Client::FOUND_ROWS

is_set = flags & check == check
```

Reply:
Very cool. So this code works regardless of the flags?

Reply:
Yup! My workaround was to use `sql.last_insert_id` which is set to 0 if an update occurs. This is not affected by the flag. However I just realized there's a bug - if the ID of the row being inserted is 0 then we would also throw the error. I've pushed a fix for this in 28d00e8.

Reply:
Strangely, playing around on the mysql console I see a `rows affected` of two when an `on duplicate key update` happens. Presumably because the auto-increment counter is affected also. It's weird that that's inconsistent with how this PR works.

Reply:
That's probably because you have `CLIENT_FOUND_ROWS` set. Its a flag passed to the connection of MySQL and the _only_ thing it affects is `on duplicate key update` queries.

From the MySQL docs:

> If you specify the CLIENT_FOUND_ROWS flag to the mysql_real_connect() C API function when connecting to mysqld, the affected-rows value is 1 (not 0) if an existing row is set to its current values.

https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html

I was able to get around this by not using rows affected but rather the `last_insert_id`

Reply:
Ah wait sorry. The value `2` reflects an update occurred with new values.

Reply:
I've made the check much more thorough now to cover a few bugs:

1. When the user decrements/increments a negative to a value of zero
2. If for some reason the insert sets the ID to 0.